### DISCUSSION TITLE
How I can enable IP_TRANSPARENT socket option for listener socket?

### DISCUSSION THREAD

**Discussion by alishir** [tokio-rs/tokio#3702] [Q&A] [answered]
Hi all,

I want to implement simple transparent proxy. According to the [kernel documentation](https://www.kernel.org/doc/Documentation/networking/tproxy.txt) I should enable IP_TRANSPARENT option for listener socket. How I can enable this option?

**Comment by Darksonn** [comment, marked as answer]
You can use the [`setsockopt`][2] method in the `nix` crate for this. 
```Rust
use nix::sys::socket::setsockopt;
use nix::sys::socket::sockopt::IpTransparent;
use std::os::unix::io::AsRawFd;

let listener = ...;
let fd = listener.as_raw_fd();
let res = setsockopt(fd, IpTransparent, &true);
```

[1]: https://docs.rs/tokio/1/tokio/net/struct.TcpListener.html#impl-AsRawFd
[2]: https://docs.rs/nix/0.20/nix/sys/socket/fn.setsockopt.html

**Reply by alishir** [reply]
Thanks a lot.
